### DISCUSSION TITLE
No way to make active link on scroll in different components using .map()

### DISCUSSION THREAD

**Discussion by nicitaacom** [thebuilder/react-intersection-observer#631] [General]
No way to make active link on scroll in different components using .map()
I want make something similar to this
![image](https://github.com/thebuilder/react-intersection-observer/assets/39565703/2104185d-81e4-4d14-90a8-e542907aa9b3)

I want here ser inView - category.id
And in Category.tsx I want create ref
![image](https://github.com/thebuilder/react-intersection-observer/assets/39565703/61de55bf-7497-466b-8272-50648d5b9e74)

But its write me that I redeclarete variable
![image](https://github.com/thebuilder/react-intersection-observer/assets/39565703/5d9a91ca-c6b9-4289-8d77-47c34102c2dc)

**Comment by nicitaacom** [comment]
Here I want set inView
![image](https://github.com/thebuilder/react-intersection-observer/assets/39565703/aa90978d-eb00-4d36-991b-047ad5f60fad)

**Comment by thebuilder** [comment]
You can't use a hook inside a loop.

**Comment by nicitaacom** [comment]
> You can't use a hook inside a loop.

I know man
But how I can create something similar to 
https://user-images.githubusercontent.com/39565703/246155279-2104185d-81e4-4d14-90a8-e542907aa9b3.png

**Reply by aditi-gawli** [reply]
There's a same case I want to implement

**Reply by MForMarlon** [reply]
Hi, I know this is a few months old, but since no one replied with a solution, here is one way to solve it using your screenshot from MDN. First, the component layout:
```
<Page>
  <Links />
  <Content />
</Page>
```

Add a state in the Page component for the visible section:
`const [visibleSection, setVisibleSection] = useState('Description');`

Pass the state as a prop into Links:
`<Links visibleSection={visibleSection} />`

Inside the Links component, render the links style change using the prop:
```
<a href="#" className={visibleSection === 'Description'  ?  'active' : ''}>Description</a>
<a href="#" className={visibleSection === 'Example'  ?  'active' : ''}>Example</a>
<a href="#" className={visibleSection === 'Values'  ?  'active' : ''}>Values</a>
```

Now pass setVisibleSection as a prop into Content:
`<Content setVisibleSection={setVisibleSection} />`

Then inside Content, set up separate `useInView` hooks for each section you want to observe,
then assign the refs into the rendering part of Content when you want to trigger the style change.
```
const [descRef] = useInView({ ..., onChange: (inView) => inView && setVisibleSection('Description') })
const [exampleRef] = useInView({ ..., onChange: (inView) => inView && setVisibleSection('Example') })
const [valuesRef] = useInView({ ..., onChange: (inView) => inView && setVisibleSection('Values') })
...
<section ref={descRef} className="description-section">...</section>
<section ref={exampleRef} className="example-section">...</section>
<section ref={valuesRef} className="values-section">...</section>
```

Hope this helps.

**Reply by nicitaacom** [reply]
It would be nice if you create quick YouTube video guide
Its not necessary

You may attach github repository with minimal example

**Reply by MForMarlon** [reply]
@nicitaacom Ok, here you go: https://codesandbox.io/p/sandbox/xenodochial-paper-jftsn2?file=%2Fsrc%2FApp.js%3A5%2C4-25%2C5

I did not implement link anchoring, but that is irrelevant.

**Reply by nicitaacom** [reply]
Work not properly

![image](https://github.com/thebuilder/react-intersection-observer/assets/39565703/bcea0635-e806-4d86-829e-525e25b189b4)

**Reply by MrBarracuda** [reply]
> Hi, I know this is a few months old, but since no one replied with a solution, here is one way to solve it using your screenshot from MDN. First, the component layout:
> 
> ```
> <Page>
>   <Links />
>   <Content />
> </Page>
> ```
> 
> Add a state in the Page component for the visible section: `const [visibleSection, setVisibleSection] = useState('Description');`
> 
> Pass the state as a prop into Links: `<Links visibleSection={visibleSection} />`
> 
> Inside the Links component, render the links style change using the prop:
> 
> ```
> <a href="#" className={visibleSection === 'Description'  ?  'active' : ''}>Description</a>
> <a href="#" className={visibleSection === 'Example'  ?  'active' : ''}>Example</a>
> <a href="#" className={visibleSection === 'Values'  ?  'active' : ''}>Values</a>
> ```
> 
> Now pass setVisibleSection as a prop into Content: `<Content setVisibleSection={setVisibleSection} />`
> 
> Then inside Content, set up separate `useInView` hooks for each section you want to observe, then assign the refs into the rendering part of Content when you want to trigger the style change.
> 
> ```
> const [descRef] = useInView({ ..., onChange: (inView) => inView && setVisibleSection('Description') })
> const [exampleRef] = useInView({ ..., onChange: (inView) => inView && setVisibleSection('Example') })
> const [valuesRef] = useInView({ ..., onChange: (inView) => inView && setVisibleSection('Values') })
> ...
> <section ref={descRef} className="description-section">...</section>
> <section ref={exampleRef} className="example-section">...</section>
> <section ref={valuesRef} className="values-section">...</section>
> ```
> 
> Hope this helps.

is there any workaround to not use the hook multiple times? I have 15+ sections that needs that ref and wonder what would be the best approach to implement that

**Reply by thebuilder** [reply]
You could handle the `useInView` inside the `<Link>` component, and pass an `onChange` prop, that the `<Link>` can trigger when `inView` changes.

Then the parent component will be notified of links as they enter/leave the viewport, and can manage the state.
